server/src/main/java/org/opensearch/common/blobstore/transfer/RemoteTransferContainer.java

```diff
@@ -0,0 +1,252 @@
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.transfer;
+
+import com.jcraft.jzlib.JZlib;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.CorruptIndexException;
+import org.opensearch.common.CheckedTriFunction;
+import org.opensearch.common.SetOnce;
+import org.opensearch.common.StreamContext;
+import org.opensearch.common.blobstore.stream.write.WriteContext;
+import org.opensearch.common.blobstore.stream.write.WritePriority;
+import org.opensearch.common.blobstore.transfer.stream.OffsetRangeInputStream;
+import org.opensearch.common.blobstore.transfer.stream.ResettableCheckedInputStream;
+import org.opensearch.common.io.InputStreamContainer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Objects;
+
+/**
+ * RemoteTransferContainer is an encapsulation for managing file transfers.
+ *
+ * @opensearch.internal
+ */
+public class RemoteTransferContainer implements Closeable {
+
+    private int numberOfParts;
+    private long partSize;
+    private long lastPartSize;
+
+    private final long contentLength;
+    private final SetOnce<InputStream[]> inputStreams = new SetOnce<>();
+    private final String fileName;
+    private final String remoteFileName;
+    private final boolean failTransferIfFileExists;
+    private final WritePriority writePriority;
+    private final long expectedChecksum;
+    private final OffsetRangeInputStreamSupplier offsetRangeInputStreamSupplier;
+    private final boolean isRemoteDataIntegritySupported;
+    private final boolean areInputStreamsDecorated;
+
+    private static final Logger log = LogManager.getLogger(RemoteTransferContainer.class);
+
+    /**
+     * Construct a new RemoteTransferContainer object
+     *
+     * @param fileName                       Name of the local file
+     * @param remoteFileName                 Name of the remote file
+     * @param contentLength                  Total content length of the file to be uploaded
+     * @param failTransferIfFileExists       A boolean to determine if upload has to be failed if file exists
+     * @param writePriority                  The {@link WritePriority} of current upload
+     * @param offsetRangeInputStreamSupplier A supplier to create OffsetRangeInputStreams
+     * @param expectedChecksum               The expected checksum value for the file being uploaded. This checksum will be used for local or remote data integrity checks
+     * @param isRemoteDataIntegritySupported A boolean to signify whether the remote repository supports server side data integrity verification
+     * @param areInputStreamsDecorated       A boolean to signify whether the streams created via {@link OffsetRangeInputStreamSupplier#get} are decorated or not
```

Comment:
What it means that input streams are decorated? Why the container should care?

Reply:
This parameter was to account for encryption related scenarios where we need to do local integrity checks. I have removed it for now, we'll address this with the client side encryption PRs. CC: @vikasvb90